### PR TITLE
Add support for authentication via TLS certs to remote CIB operations

### PR DESCRIPTION
**Pull request by clumens**
The issue to be solved is adding TLS certs to Pacemaker Remote operations.  However, remote CIB operations are a more standalone area so it's easier to develop and test that out first.  What I've tested out on this:

* If certs are enabled on both ends, authentication without certs is not allowed.
* If certs are enabled on the server and some of the environment variables are missing on the client, access is denied.
* If certs are disabled on the server but the client tries to authenticate using a cert, access is denied (this could be a bug, or it could be user error).
* If certs are disabled on the server, authentication proceeds as it always has.

What remains to be done:

* [ ] Improve error reporting on the client side
* [x] Document the procedure for making your own CA, certs, and keys
* [x] Test certificate revocation lists
* [ ] Test if each end gets a certificate but it's not signed
* [ ] Figure out PKCS#11 support
* [x] Warn/log when a certificate is close to the expiration date
* [ ] Should we allow storing certs encrypted?
* [ ] `daemons/based/based_remote.c` doesn't appear to clean up any gnutls stuff on exit
* [ ] Can we deprecate remote CIB operations without a cert?

I've removed distinguished names from the todo list.  Comparing DNs seems pretty fraught - they can include unicode characters, they could potentially be in different orders, parts can be missing, etc.  Most people suggest not comparing them as strings, and I can't find any references to a library that'll do it for us.  gnutls doesn't seem to have any function for that, at least not a public one.  I guess I'd just suggest handing out one cert to each client and if something comes up, revoke that cert.

### CODE REVIEW

File: lib/common/remote.c [hunk 1]
```diff
@@ -124,6 +125,48 @@ localized_remote_header(pcmk__remote_t *remote)
...
+    }
+
+    gnutls_x509_crt_init(&cert);
+    gnutls_x509_crt_import(cert, datum, GNUTLS_X509_FMT_DER);
+
+    expiry = gnutls_x509_crt_get_expiration_time(cert);
+
+    if (expiry != -1) {
+        time_t now = time(NULL);
+        double seconds_to_expiry = difftime(expiry, now);
```
Comment (kgaillot) [inline lib/common/remote.c:155]: I'd rather just check `expiry - now`

File: doc/sphinx/Pacemaker_Administration/configuring.rst [hunk 2]
```diff
@@ -189,9 +189,30 @@ the cluster.
...
+
+Then, create a public/private key pair and CSR for each remote system that you
+wish to have remote access.  Use the CA to sign the CSRs.  It is recommended to
+use a unique certificate for each remote system so they can be revoked if
+necessary.
+
+The server's public/private key pair and signed certificate should be installed
+to the ``/etc/pacemaker`` directory and owned by ``CIB_user``. Remember that
+private keys should not be readable by anyone other than their owner. Finally,
+edit the ``/etc/sysconfig/pacemaker`` file to refer to these credentials:
```
Comment (kgaillot) [inline doc/sphinx/Pacemaker_Administration/configuring.rst:209]: |PCMK_CONFIG_FILE|

**Review by kgaillot — COMMENTED** [review]
awesome :)

File: etc/sysconfig/pacemaker.in [hunk 3]
```diff
@@ -198,6 +198,41 @@
...
+# If set, along with PCMK_key_file and PCMK_cert_file, X509 authentication
+# will be enabled for remote CIB connections.
+#
+# Default: PCMK_ca_file=""
+
+# PCMK_cert_file
+#
+# The location of a file containing the certificate list for the matching
+# PCMK_key_file, in PEM format.  If set, along with PCMK_ca_file and PCMK_key_file,
+# X509 authentication will be enabled for remote CIB connections.
```
[the remark this answers is not in the document]
Comment (clumens) [reply etc/sysconfig/pacemaker.in:216]: Correct - the client cert gets provided on the command line when you run cibadmin (or whatever other tool).  I'll make this more clear in the comment here.

[the remark this answers is not in the document]
Comment (kgaillot) [reply etc/sysconfig/pacemaker.in:216]: Command-line tools should use new `CIB_*` environment variables to stay consistent with the rest of remote admin

**Comment by clumens** [comment]
These patches are pretty different now - it's definitely worth another in-depth look.

File: lib/common/utils.c [hunk 4]
```diff
@@ -343,11 +343,19 @@ crm_generate_uuid(void)
...
+{
+    crm_trace("%s", msg);
+}
+
 void
 crm_gnutls_global_init(void)
 {
     signal(SIGPIPE, SIG_IGN);
     gnutls_global_init();
+    gnutls_global_set_log_level(9);
```
Comment (kgaillot) [inline lib/common/utils.c]: 9 still shows private keys, 8 would be fine (though it looks like they don't actually use 7 or 8 currently)

File: include/crm/common/tls_internal.h [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+#include <gnutls/gnutls.h>  // gnutls_session_t, gnutls_dh_params_t, etc.
+
+#include <crm/common/ipc_internal.h>        // pcmk__client_t
+#include <crm/common/remote_internal.h>     // pcmk__remote_t
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// internal functions from remote.c
```
Comment (kgaillot) [inline include/crm/common/tls_internal.h:41]: can drop this comment

File: lib/common/tls.c [hunk 6]
```diff
@@ -42,6 +42,93 @@ tls_cred_str(gnutls_credentials_type_t cred_type)
...
+     */
+    gnutls_global_init();
+    gnutls_global_set_log_level(9);
+    gnutls_global_set_log_function(_gnutls_log_func);
+
+    if (server) {
+        rc = pcmk__init_tls_dh(&(*tls)->dh_params);
+        if (rc != pcmk_rc_ok) {
+            free(*tls);
+            *tls = NULL;
```
Comment (kgaillot) [inline lib/common/tls.c:203]: also have to deinit gnutls

File: lib/lrmd/lrmd_client.c [hunk 7]
```diff
@@ -653,8 +653,7 @@ lrmd_tls_connection_destroy(gpointer userdata)
     native->remote->start_state = NULL;
     native->source = 0;
     native->sock = 0;
-    native->psk_cred_c = NULL;
-    native->sock = 0;
+    native->tls = NULL;
```
Comment (kgaillot) [inline lib/lrmd/lrmd_client.c:656]: not necessary, you nulled it earlier

File: include/crm/common/util_compat.h [hunk 8]
```diff
@@ -26,6 +26,9 @@ extern "C" {
  *             release.
  */
 
+//! \deprecated Use pcmk__init_tls() instead
```
Comment (kgaillot) [inline include/crm/common/util_compat.h:29]: "Use gnutls_global_init() instead" (internal APIs aren't public)

File: lib/common/tls.c [hunk 9]
```diff
@@ -218,10 +218,16 @@ pcmk__init_tls(pcmk__tls_t **tls, bool server, gnutls_credentials_type_t cred_ty
...
-        (*tls)->key_file = pcmk__env_option(PCMK__ENV_KEY_FILE);
+        if (server) {
+            (*tls)->ca_file = pcmk__env_option(PCMK__ENV_CA_FILE);
+            (*tls)->cert_file = pcmk__env_option(PCMK__ENV_CERT_FILE);
+            (*tls)->crl_file = pcmk__env_option(PCMK__ENV_CRL_FILE);
+            (*tls)->key_file = pcmk__env_option(PCMK__ENV_KEY_FILE);
+        } else {
+            (*tls)->ca_file = getenv("CIB_ca_file");
+            (*tls)->cert_file = getenv("CIB_cert_file");
+            (*tls)->key_file = getenv("CIB_key_file");
```
Comment (kgaillot) [inline lib/common/tls.c:229]: doesn't the client need a crl too to ensure the server cert is valid?

File: doc/sphinx/Pacemaker_Administration/configuring.rst [hunk 10]
```diff
@@ -204,6 +225,14 @@ cluster, set the following environment variables:
 
 Only the Pacemaker daemon user (|CRM_DAEMON_USER|) may be used as ``CIB_user``.
 
+To use TLS certificates, the administrator's machine also needs their
+public/private key pair, signed client certificate, and root CA certificate.
+Those must additionally be specified with the following environment variables:
+
+* :ref:`PCMK_ca_file <file>`
+* :ref:`PCMK_cert_file <file>`
+* :ref:`PCMK_key_file <file>`
```
Comment (kgaillot) [inline doc/sphinx/Pacemaker_Administration/configuring.rst:234]: s/PCMK/CIB/

**Review by kgaillot — COMMENTED** [review]
this is awesome

**Comment by clumens** [comment]
Added [96eb403](https://github.com/ClusterLabs/pacemaker/pull/3738/commits/96eb4036f95fdbbc7d96f249949f35de24dd73ab) which came up as a result of a coverity complaint, and that caused some changes later on due to types.  I also enabled CRLs for clients.

**Comment by kgaillot** [comment]
Normally I wouldn't backport any new features after rc1 is released, but I think this would be a good one to do because 3.0.0 is a new major release, and rc1 was released deliberately earlier than it was ready so that people could start experimenting; and because remote CIB admin is a well-isolated feature, so this shouldn't pose any risks for the cluster itself.

**Comment by kgaillot** [comment]
BTW I didn't think of this with the initial review, but is CIB_user still used to authenticate the user, on top of the certificates authenticating the host (and encrypting the communication)?

**Comment by clumens** [comment]
> BTW I didn't think to check this with the initial review, but is PAM still used to authenticate CIB_user, on top of the certificates authenticating the host (and encrypting the communication)?

That's correct - you still need to provide a password and set CIB_user (though I think we still default to the cluster user if that's not set).

**Pull request merged**